### ISSUE TITLE
NextJS Error - does not work using workspaces

### ISSUE THREAD

**Issue by callumthomson** [denoland/deno#26125]
Version: Deno 2.0.0

# Summary
Installing NextJS works, and commands work normally, but when Next is made part of a workspace, running any commands (e.g. `dev`, `build`) all fail. Running on macOS.

# Error
```
deno task build
Task build next build
  ▲ Next.js 14.2.15

   Creating an optimized production build ...
<w> [webpack.cache.PackFileCacheStrategy] Caching failed for pack: Error: Can't resolve 'busboy' in '/Users/callum/Library/Caches/deno/npm/registry.npmjs.org/next/14.2.15'
<w> while resolving 'busboy' in /Users/callum/Library/Caches/deno/npm/registry.npmjs.org/next/14.2.15 to a directory
Failed to compile.

../../../../Library/Caches/deno/npm/registry.npmjs.org/next/14.2.15/dist/build/webpack/loaders/next-route-loader/index.js?kind=PAGES&page=%2F_error&preferredRegion=&absolutePagePath=next%2Fdist%2Fpages%2F_error&absoluteAppPath=next%2Fdist%2Fpages%2F_app&absoluteDocumentPath=next%2Fdist%2Fpages%2F_document&middlewareConfigBase64=e30%3D!
Error: Invariant: Expected relative import to start with "next/", found "14.2.15/dist/server/future/route-modules/pages/module.compiled"
    at file:///Users/callum/Library/Caches/deno/npm/registry.npmjs.org/next/14.2.15/dist/build/load-entrypoint.js:36:19
    at RegExp.[Symbol.replace] (<anonymous>)
    at String.replaceAll (<anonymous>)
    at loadEntrypoint (file:///Users/callum/Library/Caches/deno/npm/registry.npmjs.org/next/14.2.15/dist/build/load-entrypoint.js:28:17)
    at Object.runMicrotasks (ext:core/01_core.js:672:26)
    at Array.processTicksAndRejections (ext:deno_node/_next_tick.ts:57:10)
    at eventLoopTick (ext:core/01_core.js:172:29)
    at async loadPages (file:///Users/callum/Library/Caches/deno/npm/registry.npmjs.org/next/14.2.15/dist/build/webpack/loaders/next-route-loader/index.js:73:16)
    at async Object.loader (file:///Users/callum/Library/Caches/deno/npm/registry.npmjs.org/next/14.2.15/dist/build/webpack/loaders/next-route-loader/index.js:115:24)

next/dist/client/components/not-found-error
Invariant: Expected relative import to start with "next/", found "14.2.15/dist/server/future/route-modules/app-page/module.compiled"

src/app/favicon.ico
Invariant: Expected relative import to start with "next/", found "14.2.15/dist/server/future/route-modules/app-route/module.compiled"

src/app/page.tsx
Invariant: Expected relative import to start with "next/", found "14.2.15/dist/server/future/route-modules/app-page/module.compiled"


> Build failed because of webpack errors
```

# Reproduction
Install NextJS:
```bash
$ mkdir deno-next-workspace
$ cd deno-next-workspace
$ deno run -A npm:create-next-app@latest # install into 'next' directory
```
See that a regular build works:
```bash
$ cd next
$ deno task build # should see that build is successful
```
Convert into a workspace by creating `deno.json` in the `deno-next-workspace` directory:
```bash
$ cd ../
$ touch deno.json # file contents below
```
```json
{
    "workspace": [
        "./next"
    ]
}
```
Run the build again and observe the error
```bash
$ cd next
$ deno task build # should see that build is unsuccessful
```

This is my firs time using Deno. Have been fighting with this issue for a bit today but still very impressed with what I've seen so far and looking forward to being able to use workspaces properly with my projects! Thanks to the team great work 🙏

**Comment by nathanwhit** [comment]
Hello! It looks like the issue is that when a package.json is in the _root_ of a workspace, we enable the `node_modules` folder by default, but when you create the workspace the package.json is no longer in the root, so we use the global cache (no node_modules folder).

TLDR;

Try adding

```
    "nodeModulesDir": "manual"
```

to your `deno.json`. Then make sure to run `deno install`, and try again.

---

Thanks for raising this! It's definitely a rough edge we should improve.

**Comment by callumthomson** [comment]
Thanks for your reply @nathanwhit 

Adding that extra line to my `deno.json` does allow the build to progress further, but does now seem to be failing on something else.

```
deno task build
Task build next build
  ▲ Next.js 14.2.15

   Creating an optimized production build ...
 ✓ Compiled successfully
 ✓ Linting and checking validity of types    
 ✓ Collecting page data    
   Generating static pages (0/5)  [    ]TypeError: Cannot read properties of null (reading 'useContext')
    at exports.useContext (file:///Users/callum/repos/callumthomson/chassis/apps/web/node_modules/react/cjs/react.production.min.js:24:495)
    at h (file:///Users/callum/repos/callumthomson/chassis/apps/web/.next/server/pages/_error.js:1:6226)
    at Wc (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:68:44)
    at Zc (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:70:253)
    at Z (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:76:89)
    at $c (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:78:98)
    at bd (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:77:404)
    at Z (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:76:217)
    at $c (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:78:98)
    at Zc (file:///Users/callum/repos/callumthomson/chassis/node_modules/.deno/react-dom@18.3.1/node_modules/react-dom/cjs/react-dom-server.browser.production.min.js:71:145)

Error occurred prerendering page "/500". Read more: https://nextjs.org/docs/messages/prerender-error

#
# more of the same ...
#

 ✓ Generating static pages (5/5)
   Generating static pages (4/5)  [    ]
> Export encountered errors on following paths:
        /_error: /404
        /_error: /500
        /_not-found/page: /_not-found
        /page: /
```

Thanks

**Comment by nathanwhit** [comment]
Ah, you've hit another unfortunate case. Try going into your package, and removing the node_modules folder. So run something like

```
cd apps/web
rm -rf node_modules
```

The node_modules folder should be only at the root of the workspace (what's happening is that deno is using the `node_modules` in `chassis`, but `nextjs` is using `apps/web/node_modules`, and the mismatch is causing issues).

**Comment by callumthomson** [comment]
That worked! I am able to complete `deno task build` now! Thanks 🙏  I guess we can consider this resolved now, although I think it's worth mentioning that, while the build succeeds, I'm getting complaints from my editor (both VSCode and WebStorm).

Exactly the same as #25359 and #24451. I've tried the solution mentioned there (`const Image = NextImage.default;`) and they actually make my editor happy but then cause another error during build.

```ts
import React from 'react';
import Image from 'next/image';

export default function Home() {
  return (
    <div className='grid grid-rows-[20px_1fr_20px] items-center justify-items-center min-h-screen p-8 pb-20 gap-16 sm:p-20 font-[family-name:var(--font-geist-sans)]'>
      <main className='flex flex-col gap-8 row-start-2 items-center sm:items-start'>
        <Image // here
          className='dark:invert'
          src='https://nextjs.org/icons/next.svg'
          alt='Next.js logo'
          width={180}
          height={38}
          priority
        />
// ...
```
```
Deno: Image cannot be used as a JSX component.
Its type
typeof import('file:/// Users/ callum/ repos/ callumthomson/ chassis/ node_modules/.deno/ next@14.2.15/ node_modules/ next/ image')
is not a valid JSX element type.
Deno: JSX element type Image does not have any construct or call signatures.
```

Updating my code to the following removes the editor errors:
```ts
import React from 'react';
import NextImage from 'next/image';

const Image = NextImage.default;

export default function Home() {
  return (
    <div className='grid grid-rows-[20px_1fr_20px] items-center justify-items-center min-h-screen p-8 pb-20 gap-16 sm:p-20 font-[family-name:var(--font-geist-sans)]'>
      <main className='flex flex-col gap-8 row-start-2 items-center sm:items-start'>
        <Image // here
          className='dark:invert'
          src='https://nextjs.org/icons/next.svg'
          alt='Next.js logo'
          width={180}
          height={38}
          priority
        />
// ...
```
But then fails to build:
```
deno task build
Task build next build
  ▲ Next.js 14.2.15

   Creating an optimized production build ...
 ✓ Compiled successfully
   Linting and checking validity of types  .Cannot read properties of undefined (reading 'bold')
```

🥲

**Comment by nathanwhit** [comment]
Ah yeah, that is also an unfortunate situation😅. Currently your best option is likely to disable the deno LSP extension for now, for instance in vscode adding `"deno.enable": false` to your `.vscode/settings.json`. NextJS's build tooling doesn't yet understand deno specific features, and next relies on "bundler" resolution, which we don't support in the LSP.

**Comment by callumthomson** [comment]
Hmm that is a shame. Does this mean that Next essentially isn't fully compatible with Deno yet?

**Comment by lucacasonato** [comment]
Closing this as the original issue is resolved. We are tracking Next.js support in LSP here: https://github.com/denoland/deno/issues/25359
